pkg/reconciler/workload/package.go

```diff
@@ -0,0 +1,122 @@
+	"context"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	workloadv1alpha1 "github.com/crossplane/crossplane/apis/workload/v1alpha1"
+)
+
+const (
+	errWrapInKubeApp = "unable to wrap object in KubernetesApplication"
+)
+
+// A Packager is responsible for packaging workloads into other objects.
+type Packager interface {
+	Package(context.Context, Workload, ...PackageWrapper) (runtime.Object, error)
+}
+
+// A PackageFn packages a workload into an object.
+type PackageFn func(context.Context, Workload) (runtime.Object, error)
+
+// Package workload into object and apply all wrappers.
+func (fn PackageFn) Package(ctx context.Context, w Workload, p ...PackageWrapper) (runtime.Object, error) {
+	obj, err := fn(ctx, w)
+	if err != nil {
+		return nil, err
+	}
+	for _, wrap := range p {
+		if obj, err = wrap(ctx, w, obj); err != nil {
+			return nil, err
+		}
+	}
+	return obj, nil
+}
+
+var _ Packager = PackageFn(NoopPackage)
+
+// NoopPackage does not package the workload and does not return error.
+func NoopPackage(ctx context.Context, w Workload) (runtime.Object, error) {
+	return nil, nil
+}
+
+// A PackageWrapper wraps the output of a workload package in another object.
+type PackageWrapper func(context.Context, Workload, runtime.Object) (runtime.Object, error)
+
+var _ PackageWrapper = NoopWrapper
+
+// NoopWrapper does not wrap the workload package and does not return error.
+func NoopWrapper(ctx context.Context, w Workload, obj runtime.Object) (runtime.Object, error) {
+	return obj, nil
+}
+
+var _ PackageWrapper = KubeAppWrapper
+
+// KubeAppWrapper wraps a packaged object in a KubernetesApplication.
+func KubeAppWrapper(ctx context.Context, w Workload, obj runtime.Object) (runtime.Object, error) {
+	if obj == nil {
+		return nil, errors.New(errWrapInKubeApp)
+	}
+	u, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	app := &workloadv1alpha1.KubernetesApplication{}
+
+	kart := &workloadv1alpha1.KubernetesApplicationResourceTemplate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: w.GetName(),
+			Labels: map[string]string{
+				labelKey: string(w.GetUID()),
+			},
+		},
+		Spec: workloadv1alpha1.KubernetesApplicationResourceSpec{
+			Template: &unstructured.Unstructured{Object: u},
+		},
+	}
+
+	app.Spec.ResourceTemplates = append(app.Spec.ResourceTemplates, *kart)
+	// A workload's package must have the same name and namespace, and must be
+	// controlled by the same owner.
+	c := true
+	meta.AddOwnerReference(app, metav1.OwnerReference{
+		APIVersion:         w.GetObjectKind().GroupVersionKind().GroupVersion().String(),
+		Kind:               w.GetObjectKind().GroupVersionKind().Kind,
+		Name:               w.GetName(),
+		UID:                w.GetUID(),
+		Controller:         &c,
+		BlockOwnerDeletion: &c,
+	})
```

Comment:
https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#NewControllerRef may help here.

Reply:
Ah yes thank you! I was looking for something like this :)